xcm/xcm-emulator/example/src/lib.rs

```diff
@@ -0,0 +1,438 @@
+				kusama_runtime::RuntimeOrigin::root(),
+				Some(3)
+			));
+			let _ = kusama_runtime::Balances::deposit_creating(
+				&ParaId::from(1).into_account_truncating(),
+				1_000_000_000_000,
+			);
+		});
+
+		let remark = kusama_runtime::RuntimeCall::System(frame_system::Call::<
+			kusama_runtime::Runtime,
+		>::remark_with_event {
+			remark: "Hello from Pumpkin!".as_bytes().to_vec(),
+		});
+		ParachainA::execute_with(|| {
+			assert_ok!(test_runtime::PolkadotXcm::force_default_xcm_version(
+				test_runtime::RuntimeOrigin::root(),
+				Some(3)
+			));
+			assert_ok!(test_runtime::PolkadotXcm::send_xcm(
+				Here,
+				Parent,
+				Xcm(vec![
+					UnpaidExecution { weight_limit: Unlimited, check_origin: None },
+					Transact {
+						origin_kind: OriginKind::SovereignAccount,
+						require_weight_at_most: Weight::from_parts(
+							INITIAL_BALANCE as u64,
+							1024 * 1024
+						),
+						call: remark.encode().into(),
+					}
+				]),
+			));
+		});
+
+		KusamaNet::execute_with(|| {
+			use kusama_runtime::{RuntimeEvent, System};
+			// TODO: https://github.com/paritytech/polkadot/pull/6824 or change this call to
+			// force_create_assets like we do in cumulus integration tests.
+			// assert!(System::events().iter().any(|r| matches!(
+			// 	r.event,
+			// 	RuntimeEvent::System(frame_system::Event::Remarked { sender: _, hash: _ })
+			// )));
+			assert!(System::events().iter().any(|r| matches!(
+				r.event,
+				RuntimeEvent::Ump(polkadot_runtime_parachains::ump::Event::ExecutedUpward(
+					_,
+					Outcome::Incomplete(_, XcmError::NoPermission)
+				))
+			)));
+		});
+	}
+
+	#[test]
+	fn xcmp() {
+		Network::reset();
+
+		let remark = test_runtime::RuntimeCall::System(
+			frame_system::Call::<test_runtime::Runtime>::remark_with_event {
+				remark: "Hello from Pumpkin!".as_bytes().to_vec(),
+			},
+		);
+		ParachainA::execute_with(|| {
+			assert_ok!(test_runtime::PolkadotXcm::send_xcm(
+				Here,
+				MultiLocation::new(1, X1(Parachain(2))),
+				Xcm(vec![Transact {
+					origin_kind: OriginKind::SovereignAccount,
+					require_weight_at_most: 20_000_000.into(),
+					call: remark.encode().into(),
+				}]),
+			));
+		});
+
+		ParachainB::execute_with(|| {
+			use test_runtime::{RuntimeEvent, System};
+			System::events().iter().for_each(|r| println!(">>> {:?}", r.event));
+
+			assert!(System::events().iter().any(|r| matches!(
+				r.event,
+				RuntimeEvent::System(frame_system::Event::Remarked { sender: _, hash: _ })
+			)));
+		});
+	}
+
+	#[test]
+	fn xcmp_through_a_parachain() {
+		use test_runtime::{PolkadotXcm, Runtime, RuntimeCall};
+
+		Network::reset();
+
+		// The message goes through: Pumpkin --> Mushroom --> Octopus
+		let remark = RuntimeCall::System(frame_system::Call::<Runtime>::remark_with_event {
+			remark: "Hello from Pumpkin!".as_bytes().to_vec(),
+		});
+		let send_xcm_to_octopus = RuntimeCall::PolkadotXcm(pallet_xcm::Call::<Runtime>::send {
```

Comment:
Where do these names come from? Why not continue with "Parachain A/B/C/..."?

Reply:
It's a refactoring bug, messages definitely need to be fixed.